File: .spec/scripts/python/release_feature.py
#!/usr/bin/env python3
"""
release_feature.py — Finalize a spec feature release: push branch + archive
session + (optionally) switch to base branch (analogous to release-feature.ps1).

Does NOT commit — assumes the caller (spec.release agent) has already committed.

Usage:
    python release_feature.py [--stay-on-branch]
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))


def _run(cmd: list[str], cwd: str) -> tuple[int, str]:
    result = subprocess.run(cmd, capture_output=True, text=True, cwd=cwd)
    return result.returncode, (result.stdout + result.stderr).strip()


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Finalize a spec feature release."
    )
    parser.add_argument(
        "--stay-on-branch",
        action="store_true",
        help="Do not switch to the base branch after releasing"
    )
    args = parser.parse_args()

    script_dir   = Path(__file__).parent
    repo_root    = str(script_dir.parents[2])  # .spec/scripts/python → repo root
    session_file = Path(repo_root) / ".spec" / "session.json"

    if not session_file.exists():
        print("[release] No active session found. Nothing to release.", file=sys.stderr)
        sys.exit(0)

    session = json.loads(session_file.read_text(encoding="utf-8"))

    # Resolve branch_name (support both flat and nested layouts)
    branch_name = session.get("branch_name")
    if not branch_name:
        branch_name = (session.get("feature") or {}).get("branch_name")

    base_branch = (session.get("git") or {}).get("base_branch")

    if not branch_name:
        print("[release] Could not determine branch_name from session.json", file=sys.stderr)
        sys.exit(1)

    # Check git
    has_git = False
    try:
        subprocess.run(["git", "--version"], capture_output=True, check=False)
        rc, _ = _run(["git", "-C", repo_root, "rev-parse", "--is-inside-work-tree"], repo_root)
        has_git = (rc == 0)
    except FileNotFoundError:
        pass

    # Push to origin
    if has_git:
        rc, remote_url = _run(["git", "-C", repo_root, "config", "--get", "remote.origin.url"], repo_root)
        if rc == 0 and remote_url.strip():
            print(f"[release] Pushing branch: {branch_name}")
            rc, out = _run(
                ["git", "-C", repo_root, "push", "origin", branch_name, "--set-upstream"],
                repo_root
            )
            if rc != 0:
                print(f"[release] Push failed; continuing with archive. ({out})", file=sys.stderr)
        else:
            print("[release] No remote 'origin' found — skipping push. Branch is available locally only.")
    else:
        print("[release] Git not available — skipping push.")

    # Archive session
    rc = subprocess.run(
        [sys.executable, str(script_dir / "manage_session.py"), "--action", "archive"]
    ).returncode
    if rc != 0:
        print("[release] Warning: archive step returned non-zero.", file=sys.stderr)

    # Switch to base branch
    if (
        not args.stay_on_branch
        and base_branch
        and base_branch != branch_name
        and has_git
    ):
        rc, _ = _run(["git", "-C", repo_root, "checkout", base_branch], repo_root)
        if rc == 0:
            print(f"[release] Switched to base branch: {base_branch}")

    base_display = base_branch or "<none>"
    print(f"[release] Done. Branch: {branch_name}  Base: {base_display}")


if __name__ == "__main__":
    main()

File: .spec/scripts/release_feature.ts
#!/usr/bin/env node
import * as fs from "node:fs";
import * as path from "node:path";
import { spawnSync } from "node:child_process";

function run(cmd: string[], cwd: string): { code: number; out: string } {
  const r = spawnSync(cmd[0], cmd.slice(1), { cwd, encoding: "utf-8" });
  return { code: r.status ?? 1, out: `${r.stdout ?? ""}${r.stderr ?? ""}`.trim() };
}

const stayOnBranch = process.argv.includes("--stay-on-branch");
const scriptDir = __dirname;
const repoRoot = path.resolve(scriptDir, "..", "..", "..");
const sessionFile = path.join(repoRoot, ".spec", "session.json");

if (!fs.existsSync(sessionFile)) {
  console.error("[release] No active session found. Nothing to release.");
  process.exit(0);
}

const session = JSON.parse(fs.readFileSync(sessionFile, "utf-8")) as {
  branch_name?: string;
  feature?: { branch_name?: string };
  git?: { base_branch?: string };
};

const branchName = session.branch_name ?? session.feature?.branch_name;
const baseBranch = session.git?.base_branch;

if (!branchName) {
  console.error("[release] Could not determine branch_name from session.json");
  process.exit(1);
}

let hasGit = false;
if (!spawnSync("git", ["--version"], { stdio: "ignore" }).error) {
  hasGit = run(["git", "-C", repoRoot, "rev-parse", "--is-inside-work-tree"], repoRoot).code === 0;
}

if (hasGit) {
  const remote = run(["git", "-C", repoRoot, "config", "--get", "remote.origin.url"], repoRoot);
  if (remote.code === 0 && remote.out.trim()) {
    console.log(`[release] Pushing branch: ${branchName}`);
    const push = run(["git", "-C", repoRoot, "push", "origin", branchName, "--set-upstream"], repoRoot);
    if (push.code !== 0) {
      console.error(`[release] Push failed; continuing with archive. (${push.out})`);
    }
  } else {
    console.log("[release] No remote 'origin' found — skipping push. Branch is available locally only.");
  }
} else {
  console.log("[release] Git not available — skipping push.");
}

const archive = spawnSync("npx", ["tsx", path.join(scriptDir, "manage_session.ts"), "--action", "archive"], {
  cwd: scriptDir,
  stdio: "inherit",
});
if ((archive.status ?? 1) !== 0) {
  console.error("[release] Warning: archive step returned non-zero.");
}

if (!stayOnBranch && baseBranch && baseBranch !== branchName && hasGit) {
  const sw = run(["git", "-C", repoRoot, "checkout", baseBranch], repoRoot);
  if (sw.code === 0) console.log(`[release] Switched to base branch: ${baseBranch}`);
}

console.log(`[release] Done. Branch: ${branchName}  Base: ${baseBranch ?? "<none>"}`);

